### PR TITLE
findmnt: implement -Q,--filter option with % comparison

### PR DESCRIPTION
**Pull request by masatake**
```
$ ./findmnt -Q 'USE% > "70%"' -oTARGET,SOURCE,FSTYPE,USE%
TARGET         SOURCE                        FSTYPE   USE%
/              /dev/mapper/fedora_dev64-root ext4      73%
/srv/sources8c /dev/loop0                    squashfs 100%
/srv/sources9c /dev/loop1                    squashfs 100%
/srv/sources7c /dev/loop2                    squashfs 100%
$ ./findmnt -Q 'FSTYPE == "nfs4"' -oTARGET,SOURCE,FSTYPE
TARGET        SOURCE                   FSTYPE
/srv/sources8 srpmix7:/var/lib/srpmix7 nfs4
$ ./findmnt -Q 'SOURCE =~ ".*loop.*"' -oTARGET,SOURCE,FSTYPE
TARGET         SOURCE     FSTYPE
/srv/sources8c /dev/loop0 squashfs
/srv/sources9c /dev/loop1 squashfs
/srv/sources7c /dev/loop2 squashfs
```

- [x] update findmnt.8
- [ ] ~update scols-filter.5~
- [x] add test cases

Though I have above TODO items. However, I would like to get comments.

I would like to know how to compare SIZE.

Edited: I withdraw the '%' comparison code from this pull request.

### CODE REVIEW

File: misc-utils/findmnt.c [hunk 1]
```diff
@@ -1216,8 +1265,12 @@ static int add_matching_lines(struct libmnt_table *tb,
 	while((fs = get_next_fs(tb, itr, findmnt))) {
 		if ((findmnt->flags & FL_TREE) || (findmnt->flags & FL_SUBMOUNTS))
 			rc = create_treenode(table, tb, fs, NULL, findmnt);
-		else
-			rc = !add_line(table, fs, NULL, findmnt);
+		else {
+			struct libscols_line *l = add_line(table, fs, NULL, findmnt);
+			if (l == filtered_line)
+				continue;
```
Comment (karelzak) [inline misc-utils/findmnt.c:1271]: Would it be better to add a "filtered" boolean argument to add_line()
```
   bool filtered = 0;
   struct libscols_line *l = add_line(table, fs, NULL, findmnt, &filtered);
   if (filtered)
     continue;
```

... rather than use a magic global variable  `filtered_line` ?

File: misc-utils/findmnt.c [hunk 2]
```diff
@@ -1880,6 +1993,10 @@ int main(int argc, char *argv[])
 			"options --target and --source can't be used together "
 			"with command line element that is not an option"));
 
+	if ((findmnt.flags & FL_POLL) && findmnt.filter)
+		errx(EXIT_FAILURE, _(
+			"options --poll and -Q,--filter can't be used together "));
+
```
Comment (karelzak) [inline misc-utils/findmnt.c:1999]: This is unnecessary. Just add `{ 'Q', 'p' }` to `ul_excl_t excl[]` .

File: misc-utils/findmnt.h [hunk 3]
```diff
@@ -35,10 +35,12 @@ enum {
 	FL_JSON		= (1 << 30),
 };
 
+struct libscols_filter;
```
Comment (karelzak) [inline misc-utils/findmnt.h:38]: Why is this necessary?

File: libsmartcols/src/filter-param.c [hunk 4]
```diff
@@ -523,6 +517,78 @@
...
+	if (rc < 0)
+		return false;
+
+	return true;
+}
+
+static int string_opers(enum filter_etype oper, struct filter_param *l,
+			struct filter_param *r, int *status)
+{
+	if ((F_EXPR_LE || F_EXPR_LT || F_EXPR_GE || F_EXPR_LT)
```
[the remark this answers is not in the document]
Comment (karelzak) [reply libsmartcols/src/filter-param.c:553]: CodeQL is right here; you want to compare the `oper` variable with F_EXPR_*, but `oper` is missing in the expression.

File: libsmartcols/src/filter-param.c [hunk 5]
```diff
@@ -110,9 +111,14 @@ struct filter_node *filter_new_param(
...
 	}
 
+	n->percentage_holder = 0;
 	if (holder == F_HOLDER_COLUMN) {
+		size_t len;
 		n->holder_name = strdup((char *) data);
 		DBG(FLTR, ul_debugobj(fltr, "new %s holder", n->holder_name));
+		len = strlen(n->holder_name);
+		if (len > 1 && n->holder_name[len - 1] == '%')
+			n->percentage_holder = 1;
```
Comment (karelzak) [inline libsmartcols/src/filter-param.c:121]: Maybe you can use
```
 n->percentage_holder = endswith(n->holder_name, "%") ? 1 : 0;
```
to keep it simple :)

File: libsmartcols/src/filter-param.c [hunk 6]
```diff
@@ -110,9 +111,14 @@ struct filter_node *filter_new_param(
 		return NULL;
 	}
 
+	n->percentage_holder = 0;
```
Comment (karelzak) [inline libsmartcols/src/filter-param.c:114]: I guess the new param is allocated by calloc(), so this is unnecessary, right?

File: libsmartcols/src/filter-param.c [hunk 7]
```diff
@@ -523,6 +517,78 @@ static int float_opers(enum filter_etype oper, struct filter_param *l,
...
+	if (!rf.empty)
+		rf.val.fnum = rfnum;
+
+	rc = float_opers(oper, &lf, &rf, status);
+	if (rc < 0)
+		return false;
+
+	return true;
+}
+
```
Comment (karelzak) [inline libsmartcols/src/filter-param.c:549]: I'm not sure about this. 

You're trying to detect percentage usage during filter evaluation, but it's obvious already during filter parse time. I'll add separate comment about it.

File: misc-utils/findmnt.h (hunk 3 above)
Comment (masatake) [reply misc-utils/findmnt.h:38]: (Fixed in locally)

File: libsmartcols/src/filter-param.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (masatake) [reply libsmartcols/src/filter-param.c:553]: (Fixed in locally.)

File: libsmartcols/src/filter-param.c [hunk 8]
```diff
@@ -419,42 +425,30 @@ static int xstrcmp(char *a, char *b)
...
-		*status = regexec(r->re, l->val.str ? : "", 0, NULL, 0) != 0;
-		break;
-	default:
-		return -EINVAL;
+	size_t len;
+	int rc;
+
+	if (p->holder_name) {
+		if (!p->percentage_holder)
+			return false;
```
[the remark this answers is not in the document]
Comment (masatake) [reply libsmartcols/src/filter-param.c:435]: (Fixed locally)

File: libsmartcols/src/filter-param.c (hunk 5 above)
Comment (masatake) [reply libsmartcols/src/filter-param.c:121]: (Fixed locally).

File: libsmartcols/src/filter-param.c (hunk 6 above)
Comment (masatake) [reply libsmartcols/src/filter-param.c:114]: (Fixed locally).

**Comment by karelzak** [comment]
I like the idea, but my suggestion for the filter implementation:

* introduce T_FLOATPERC token to libsmartcols/src/filter-scanner.l for numbers like `<int>.<int>%`
* introduce T_HOLDERPERC token for names like `USE%`
* add `enum { F_FLAG_NONE = 0, F_FLAG_PERC = (1 << 1) }` for extra flags to libsmartcols/src/smartcolsP.h. 
* extend filter_new_param() to accept flags and struct filter_param to remember the flags
* add to libsmartcols/src/filter-parser.y
```
   | T_FLOATPERC      { $$ = filter_new_param(fltr, SCOLS_DATA_FLOAT, 0,  F_FLAG_PERC, (void *) (&$1)); }
   | T_HOLDERPERC   { $$ = filter_new_param(fltr, SCOLS_DATA_FLOAT, F_HOLDER_COLUMN, F_FLAG_PERC,  (void *) $1); }
```
* The only tricky place will be fetch_holder_data(), because applications keep percentages as strings usually, so you need to be sure that data (without tailing %) are cast to SCOLS_DATA_FLOAT if F_FLAG_PERC is defined.

The result will be that all is based on floats, and all expression operators work as expected.

File: misc-utils/findmnt.c (hunk 2 above)
Comment (masatake) [reply misc-utils/findmnt.c:1999]: (Fixed locally).

File: misc-utils/findmnt.c (hunk 1 above)
Comment (masatake) [reply misc-utils/findmnt.c:1271]: (Fixed locally).

**Comment by masatake** [comment]
@karelzak Thank you for reviewing. As you wrote, handling '%' should be done in the earlier stage. I would like to withdraw the change handling '%' from this pull request. In this pull request, I would like to focus on adding -Q option to findmnt.

File: misc-utils/findmnt.c (hunk 2 above)
Comment (masatake) [reply misc-utils/findmnt.c]: After updating as you show, findmnt accepts -Q and -p:
```
[yamato@dev64]~/var/util-linux/util-linux% ./findmnt -p -Q 'FSTYPE == "xfs"'
ACTION     TARGET SOURCE    FSTYPE OPTIONS
umount     /tmp/b tmpfs[/a] tmpfs  rw,nosuid,nodev,seclabel,nr_inodes=1048576,inode64^C
```

Am I missing something?

Comment (karelzak) [reply misc-utils/findmnt.c]: See the `ul_excl_t excl[]`, there is a note ` /* rows and cols in ASCII order */`. You have to add `{ 'Q', 'p' }` after `{ 'P','l','r','x' }`.

File: misc-utils/findmnt.8.adoc [hunk 9]
```diff
@@ -131,6 +131,15 @@ Print only pseudo filesystems.
 *--shadow*::
 Print only filesystems over-mounted by another filesystem.
 
+*-Q*, *--filter* _expr_::
+Print only the filesystems that meet the conditions specified by the expr.
++
+This feature is EXPERIMENTAL. See also *scols-filter*(5).
+For exampleinclude xfs having more than 50% free space:
```
Comment (karelzak) [inline misc-utils/findmnt.8.adoc:138]: `For exampleinclude` -> "For example, include"

**Comment by karelzak** [comment]
It's a good idea to make a separate pull request for '%' support.

File: misc-utils/findmnt.c (hunk 2 above)
Comment (masatake) [reply misc-utils/findmnt.c]: Oh, I see. I must be more careful. Fixed.

File: misc-utils/findmnt.8.adoc (hunk 9 above)
Comment (masatake) [reply misc-utils/findmnt.8.adoc:138]: It looks I'm tired. I have made many poor mistakes. Fixed locally.

**Comment by masatake** [comment]
After adding a test case for `-Q`, I found the current -Q implementation drops "tree information."

In addition to the '%' issue, I would like to work on keeping the information in the next development cycle.

**Comment by karelzak** [comment]
@masatake, thanks for all the work (and sorry for the delay)!

I have fixed the tree issue with -Q (see https://github.com/util-linux/util-linux/pull/3176)
